### PR TITLE
Jakarta EE 10 - JPA 3.1 support

### PR DESCRIPTION
**Pull request by jGauravGupta**
Signed-off-by: Gaurav Gupta <gaurav.gupta@payara.fish>

This PR adds support for JPA 3.0, and JPA 3.1 to enable the visual editor for the `persistence.xml` deployment descriptor.

### CODE REVIEW

**Comment by mbien** [comment]
@jGauravGupta is this still WIP?

**Comment by jGauravGupta** [comment]
Hi @mbien ,

Yes, I will resume the work on this PR. It can probably be targeted for NB18.

**Comment by nicolaken** [comment]
Hi @jGauravGupta, what is still missing? 
I do not no much about the Netbeans codebase and I had tried to do something about this problem, but now that I see how many files you have touched, I understand why I failed.

File: java/j2ee.persistence/src/org/netbeans/modules/j2ee/persistence/provider/Provider.java [hunk 1]
```diff
@@ -83,7 +83,11 @@ public boolean isOnClassPath(ClassPath cp)
         boolean ret = cp.findResource(classRelativePath) != null;
         if(ret && version != null)
         {
-            if (Persistence.VERSION_2_2.equals(version)) {
+            if(Persistence.VERSION_3_1.equals(version)){
+                ret &= cp.findResource("jakarta/persistence/spi/TransformerException.class.class") != null;
+            } else if(Persistence.VERSION_3_0.equals(version)){
+                ret &= cp.findResource("jakarta/persistence/Entity.class") != null;
+            | else if (Persistence.VERSION_2_2.equals(version)) {
```
Comment (mbien) [inline java/j2ee.persistence/src/org/netbeans/modules/j2ee/persistence/provider/Provider.java:90]: typo `|` ->` }`

**Comment by mbien** [comment]
sounds like this would fix #3756?

File: enterprise/glassfish.javaee/src/org/netbeans/modules/glassfish/javaee/Hk2JpaSupportImpl.java [hunk 2]
```diff
@@ -97,10 +109,12 @@ private static class JpaSupportVector {
...
         for (GlassFishVersion version : GlassFishVersion.values()) {
             jpaSupport[version.ordinal()] = new JpaSupportVector(
-                    GlassFishVersion.lt(version, GlassFishVersion.GF_6),
-                    GlassFishVersion.lt(version, GlassFishVersion.GF_6) && GlassFishVersion.ge(version, GlassFishVersion.GF_3),
-                    GlassFishVersion.lt(version, GlassFishVersion.GF_6) && GlassFishVersion.ge(version, GlassFishVersion.GF_4),
-                    GlassFishVersion.lt(version, GlassFishVersion.GF_6) && GlassFishVersion.ge(version, GlassFishVersion.GF_5)
+                    GlassFishVersion.lt(version, GlassFishVersion.GF_7_0_0),
+                    GlassFishVersion.lt(version, GlassFishVersion.GF_7_0_0) && GlassFishVersion.ge(version, GlassFishVersion.GF_3),
+                    GlassFishVersion.lt(version, GlassFishVersion.GF_7_0_0) && GlassFishVersion.ge(version, GlassFishVersion.GF_4),
+                    GlassFishVersion.lt(version, GlassFishVersion.GF_7_0_0) && GlassFishVersion.ge(version, GlassFishVersion.GF_5),
```
Comment (pepness) [inline enterprise/glassfish.javaee/src/org/netbeans/modules/glassfish/javaee/Hk2JpaSupportImpl.java:115]: JPA 1.0, 2.0, 2.1 and 2.2 are not supported by GlassFish 6 and upward, so the comparison for JPA below 3.0  should be against `GlassFishVersion.GF_6`

File: java/j2ee.persistence/licenseinfo.xml [hunk 3]
```diff
@@ -71,6 +71,13 @@
         </license>
         <comment type="CATEGORY_B">Used for XML Validation and code completion at runtime.</comment>
     </fileset>
+    <fileset>
+        <file>src/org/netbeans/modules/j2ee/persistence/dd/resources/orm_3_1.xsd</file>
```
Comment (pepness) [inline java/j2ee.persistence/licenseinfo.xml:75]: orm 3.1 is already in this file, see line 67

**Comment by pepness** [comment]
I think this PR is ready. When using JPA 3.0 or 3.1 the namespace will be `jakarta.persistence.`; if  not it will use `javax.persistence.`
JPA 3.0 with EclipseLink provider
Screenshot from 2023-03-19 13-22-09
JPA 2.2 with OpenJPA provider
Screenshot from 2023-03-19 13-22-42

**Review by pepness — APPROVED** [review]

**Comment by matthiasblaesing** [comment]
@pepness @jGauravGupta it would be good to get this merged shortly, could two please check if you want to cleanup the commits or merge?

**Comment by matthiasblaesing** [comment]
This is marked as reviewed, I eyeballed it today and it looked sane, so I'll merge this sometime next weekend before branch of NB18. If someone wants to intervene, now would be a good time.

**Comment by matthiasblaesing** [comment]
Lets get this in.

**Pull request merged**